From f1448ffaff4679c54b6140f78a4d9e8fdf2c6948 Mon Sep 17 00:00:00 2001
From: Christian Thaler <christian.thaler@tes-dst.com>
Date: Wed, 22 Feb 2017 10:54:08 +0100
Subject: [PATCH] DW_MMC: DT prop to select power enable polarity.

Due to a HW bug in Arria 10 SX 660 ES (see errata) (inverted
power enable signal), the power enable signal's polarity can
now be selected.
---
 drivers/mmc/host/dw_mmc.c  | 25 +++++++++++++++++++++++--
 include/linux/mmc/dw_mmc.h |  2 ++
 2 files changed, 25 insertions(+), 2 deletions(-)

diff --git a/drivers/mmc/host/dw_mmc.c b/drivers/mmc/host/dw_mmc.c
index 7ed9148..d3bcfe2 100644
--- a/drivers/mmc/host/dw_mmc.c
+++ b/drivers/mmc/host/dw_mmc.c
@@ -1201,7 +1201,12 @@ static void dw_mci_set_ios(struct mmc_host *mmc, struct mmc_ios *ios)
 		}
 		set_bit(DW_MMC_CARD_NEED_INIT, &slot->flags);
 		regs = mci_readl(slot->host, PWREN);
-		regs |= (1 << slot->id);
+		if(slot->host->pdata->pwr_en) {
+			regs |= (1 << slot->id);
+		}
+		else {
+			regs &= ~(1 << slot->id);
+		}
 		mci_writel(slot->host, PWREN, regs);
 		break;
 	case MMC_POWER_ON:
@@ -1240,7 +1245,12 @@ static void dw_mci_set_ios(struct mmc_host *mmc, struct mmc_ios *ios)
 		slot->host->vqmmc_enabled = false;
 
 		regs = mci_readl(slot->host, PWREN);
-		regs &= ~(1 << slot->id);
+		if(slot->host->pdata->pwr_en) {
+			regs &= ~(1 << slot->id);
+		}
+		else {
+			regs |= (1 << slot->id);
+		}
 		mci_writel(slot->host, PWREN, regs);
 		break;
 	default:
@@ -2673,6 +2683,17 @@ static struct dw_mci_board *dw_mci_parse_dt(struct dw_mci *host)
 	if (!pdata)
 		return ERR_PTR(-ENOMEM);
 
+	/* find out if we have a positive or negative power
+	 * enable (Arria 10 SX 660 ES errata) */
+	if (of_property_read_u32(dev->of_node, "pwr-en",
+				&pdata->pwr_en)) {
+		dev_info(dev, "pwr-en property not found, "
+				"assuming positive power enable signal. "
+				"(must be negative for Arria 10 SX 660 "
+				"ES due to HW bug)\n");
+		pdata->pwr_en = 1;
+	}
+
 	/* find out number of slots supported */
 	if (of_property_read_u32(dev->of_node, "num-slots",
 				&pdata->num_slots)) {
diff --git a/include/linux/mmc/dw_mmc.h b/include/linux/mmc/dw_mmc.h
index 1211199..c6e1b40 100644
--- a/include/linux/mmc/dw_mmc.h
+++ b/include/linux/mmc/dw_mmc.h
@@ -247,6 +247,8 @@ struct block_settings {
 struct dw_mci_board {
 	u32 num_slots;
 
+	u32 pwr_en; /* Power enable signal polarity (Arria 10 SX 660 ES errata)*/
+
 	u32 quirks; /* Workaround / Quirk flags */
 	unsigned int bus_hz; /* Clock speed at the cclk_in pad */
 
-- 
1.9.1

